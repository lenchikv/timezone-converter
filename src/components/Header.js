import React from 'react';
import AppBar from '@material-ui/core/AppBar';

import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import CompareArrows from '@material-ui/icons/CompareArrows';

const Header = () => (
  <div>
    <div className="appbarwrapper">
      <AppBar position="static">
        <Toolbar>
          <IconButton className="iconbuttonsyle" color="inherit" aria-label="Menu">
            <CompareArrows />
          </IconButton>
          <Typography variant="title" color="inherit" className="headertypoclass">
            React Timezone Converter
          </Typography>
        </Toolbar>
      </AppBar>
    </div>
  </div>
);

export default Header;
