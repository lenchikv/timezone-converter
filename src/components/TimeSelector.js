import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FormControl from '@material-ui/core/FormControl';
import TextField from '@material-ui/core/TextField';
import 'moment-timezone';
import { connect } from 'react-redux';
import { setDate, updateBasisDate } from '../store/actions';
import { setDateToFormatted, setDefaultTime, getDefaultTimezone } from '../helpers/DateHelper';

class TimeSelector extends Component {
  constructor(props) {
    super(props);
    const { updateDate } = this.props;
    this.defaultTime = setDefaultTime();
    updateDate({
      basisDate: setDateToFormatted(),
      currentTimeZone: getDefaultTimezone(),
    });
  }

  addZoneToList = zone => {
    const { zoneList = [] } = this.state;
    if (!zoneList.includes(zone)) {
      this.setState(state => {
        const newList = state.zoneList.concat(zone);
        return { zoneList: newList };
      });
    }
  };

  setNewTime = e => {
    e.preventDefault();
    const newTime = e.target.value.split(':');
    if (!Array.isArray(newTime) || newTime.length !== 2) {
      return;
    }
    const { operationDate, updateBasis } = this.props;
    updateBasis({ basisDate: setDateToFormatted(operationDate, newTime) });
  };

  render() {
    return (
      <FormControl>
        <TextField
          id="time"
          label="Time"
          placeholder="Time"
          className="contact-page-name-list-field"
          style={{ width: 200 }}
          type="time"
          defaultValue={this.defaultTime}
          inputProps={{ step: 300 }}
          onChange={this.setNewTime}
        />
      </FormControl>
    );
  }
}

TimeSelector.propTypes = {
  operationDate: PropTypes.string,
  updateDate: PropTypes.func.isRequired,
  updateBasis: PropTypes.func.isRequired,
};

TimeSelector.defaultProps = {
  operationDate: '',
};

const mapDispatchToProps = dispatch => ({
  updateDate: date => dispatch(setDate(date)),
  updateBasis: date => dispatch(updateBasisDate(date)),
});

const mapStateToProps = state => ({ operationDate: state.date.basisDate });

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TimeSelector);
