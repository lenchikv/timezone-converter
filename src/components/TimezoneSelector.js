import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { getTimezoneList } from '../helpers/DateHelper';
import { setTimezones } from '../store/actions';

// Additional export for testing redux by jest
export class TimezoneSelector extends Component {
  constructor(props) {
    super(props);
    const defaultTimezone = 'Etc/GMT';
    const { updateTimezones } = this.props;
    updateTimezones([defaultTimezone]);
    const timezonesMomentList = getTimezoneList();
    this.state = {
      timezone: defaultTimezone,
      timezonesMomentList,
    };
  }

  handleZoneChange = e => {
    const timezone = e.target.value;
    const { timezones, updateTimezones } = this.props;

    if (!timezones.includes(timezone)) {
      timezones.push(timezone);
      updateTimezones(timezones);
      const { timezone: oldTimezone } = this.state;
      if (!oldTimezone || oldTimezone !== timezone) {
        this.setState({ timezone });
      }
    }
  };

  render() {
    const { timezone, timezonesMomentList } = this.state;
    return (
      <FormControl className="form-styling" id="timezones-list">
        <InputLabel htmlFor="timezone">Add Timezone</InputLabel>
        <Select
          id="timezone"
          style={{ width: 200 }}
          value={timezone}
          onChange={this.handleZoneChange}
        >
          {timezonesMomentList.map(zone => (
            <MenuItem key={zone} value={zone}>
              {zone.replace('Etc/', '')}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  updateTimezones: list => dispatch(setTimezones(list)),
});

const mapStateToProps = state => ({
  timezones: Array.isArray(state.timezones) ? state.timezones : [],
});

TimezoneSelector.propTypes = {
  timezones: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Object)]).isRequired,
  updateTimezones: PropTypes.func.isRequired,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TimezoneSelector);
