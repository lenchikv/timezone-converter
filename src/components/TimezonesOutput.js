import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'moment-timezone';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import HighlightOff from '@material-ui/icons/HighlightOff';
import Schedule from '@material-ui/icons/Schedule';
import IconButton from '@material-ui/core/IconButton';
import Avatar from '@material-ui/core/Avatar';
import { connect } from 'react-redux';
import { setTimezones } from '../store/actions';
import { getMomentDate, getTimezonedDate } from '../helpers/DateHelper';

// Additional export for testing
export class TimezonesOutput extends Component {
  removeTimezoneFromList = timezone => {
    const { updateTimezones, timezones } = this.props;
    const newTimezones = timezones.filter(item => item !== timezone);
    updateTimezones(newTimezones);
  };

  render() {
    const { operationDate, timezones } = this.props;
    const momentDate = getMomentDate(operationDate);
    return (
      <List style={{ width: 250, margin: 'auto' }}>
        {timezones.map((timezone, index) => (
          <ListItem key={timezone}>
            <ListItemAvatar>
              <Avatar>
                <Schedule />
              </Avatar>
            </ListItemAvatar>
            <ListItemText
              primary={getTimezonedDate(momentDate, timezone)}
              secondary={timezone.replace('Etc/', '')}
            />
            {index !== 0 && (
              <ListItemSecondaryAction onClick={e => this.removeTimezoneFromList(timezone)}>
                <IconButton aria-label="Delete">
                  <HighlightOff />
                </IconButton>
              </ListItemSecondaryAction>
            )}
          </ListItem>
        ))}
      </List>
    );
  }
}

TimezonesOutput.propTypes = {
  timezones: PropTypes.arrayOf(PropTypes.string).isRequired,
  operationDate: PropTypes.string.isRequired,
  updateTimezones: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  updateTimezones: list => dispatch(setTimezones(list)),
});

const mapStateToProps = state => ({
  operationDate: state.date.basisDate,
  timezones: state.timezones,
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TimezonesOutput);
