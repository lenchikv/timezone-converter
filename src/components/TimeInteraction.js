import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import TimeSelector from './TimeSelector';
import TimezoneSelector from './TimezoneSelector'; // eslint-disable-line
import { getDefaultTimezone } from '../helpers/DateHelper';

const TimeInteraction = () => {
  const currentTimeZone = getDefaultTimezone();

  return (
    <Grid item xs={12} lg={6} md={6} className="grid-styling">
      <Typography variant="subheading" gutterBottom className="typography-styling">
        Local Timezone: {currentTimeZone}
      </Typography>
      <Grid item xs={12} lg={12} md={12}>
        <TimeSelector />
      </Grid>
      <Grid item xs={12} lg={12} md={12}>
        <TimezoneSelector />
      </Grid>
    </Grid>
  );
};

export default TimeInteraction;
