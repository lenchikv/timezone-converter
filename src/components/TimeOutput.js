import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import TimezonesOutput from './TimezonesOutput'; // eslint-disable-line

const TimeOutput = () => (
  <Grid item xs={12} lg={6} md={6} style={{ marginTop: 10 }}>
    <Typography variant="subheading" gutterBottom style={{ padding: 15 }}>
      Converted Timezones
    </Typography>
    <TimezonesOutput />
  </Grid>
);

export default TimeOutput;
