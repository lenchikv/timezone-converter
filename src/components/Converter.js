import React from 'react';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import TimeInteraction from './TimeInteraction';
import TimeOutput from './TimeOutput';

const Converter = () => (
  <div className="contact-page-wrapper">
    <Grid container spacing={24} justify="center">
      <Grid item xs={12} md={10} className="grid-styling">
        <Paper className="contact-page-paper">
          <Typography variant="display1" gutterBottom className="typography-styling">
            Converter
          </Typography>
          <Grid container spacing={24}>
            <TimeInteraction />
            <TimeOutput />
          </Grid>
        </Paper>
      </Grid>
    </Grid>
  </div>
);

export default Converter;
