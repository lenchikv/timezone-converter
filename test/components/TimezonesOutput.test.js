import { shallow } from 'enzyme';
import React from 'react';
import { TimezonesOutput } from '../../src/components/TimezonesOutput';

describe('TimezonesOutput', () => {
  it('renders correctly', () => {
    const operationDate = '2019-01-24T15:18:22+02:00';
    const timezones = ['Etc/Greenwich', 'Etc/UTC', 'Etc/GMT+6'];
    const tree = shallow(<TimezonesOutput operationDate={operationDate} timezones={timezones} />);
    expect(tree).toMatchSnapshot();
  });
});
