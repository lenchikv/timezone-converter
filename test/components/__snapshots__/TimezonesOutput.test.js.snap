// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimezonesOutput renders correctly 1`] = `
<WithStyles(List)
  style={
    Object {
      "margin": "auto",
      "width": 250,
    }
  }
>
  <WithStyles(ListItem)
    key="Etc/Greenwich"
  >
    <WithStyles(ListItemAvatar)>
      <WithStyles(Avatar)>
        <pure(Schedule) />
      </WithStyles(Avatar)>
    </WithStyles(ListItemAvatar)>
    <WithStyles(ListItemText)
      primary="1:18 pm"
      secondary="Greenwich"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    key="Etc/UTC"
  >
    <WithStyles(ListItemAvatar)>
      <WithStyles(Avatar)>
        <pure(Schedule) />
      </WithStyles(Avatar)>
    </WithStyles(ListItemAvatar)>
    <WithStyles(ListItemText)
      primary="1:18 pm"
      secondary="UTC"
    />
    <WithStyles(ListItemSecondaryAction)
      onClick={[Function]}
    >
      <WithStyles(IconButton)
        aria-label="Delete"
      >
        <pure(HighlightOff) />
      </WithStyles(IconButton)>
    </WithStyles(ListItemSecondaryAction)>
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    key="Etc/GMT+6"
  >
    <WithStyles(ListItemAvatar)>
      <WithStyles(Avatar)>
        <pure(Schedule) />
      </WithStyles(Avatar)>
    </WithStyles(ListItemAvatar)>
    <WithStyles(ListItemText)
      primary="7:18 am"
      secondary="GMT+6"
    />
    <WithStyles(ListItemSecondaryAction)
      onClick={[Function]}
    >
      <WithStyles(IconButton)
        aria-label="Delete"
      >
        <pure(HighlightOff) />
      </WithStyles(IconButton)>
    </WithStyles(ListItemSecondaryAction)>
  </WithStyles(ListItem)>
</WithStyles(List)>
`;
